### ISSUE TITLE
Virtual fields of parent features are not updated after inserting/updating child features

### ISSUE THREAD

**Issue by koenigherford** [qgis/QGIS#41243]
As described in #41233 and #41239 sometimes it’s required that every parent feature has one or more child features and additionally they have to meet constraints too.

To be able to enforce that „the count of child features is always>0“) you can try a little workaround:
Usually a relation in QGIS is established by using the "id"-field of the parent-layer as the referencing field in the child-layer.
This has the effect, that a parent-feature has to be saved before child-features can be added (because the `id` has the value `nextval('mysequence_seq'::regclass) `before saving, which is not accepted by QGIS as valid referencing field-value.
In this case a parent feature will be saved without a child feature. This may not be a accepted scenario when collecting data if it is required, that a parent has at least one child every time. (Additionally this causes unnecessary efforts of several additional steps to save/identify the same parent feature and add child features after that.) 
As a workaround you can set up a auxilliary-id-field with a hidden, not unique default-value (in my case by using timestamp-data). With this referenced field you will be able to add child-features without saving the parent first (when #40129 is fixed - because default aux-id's are not evaluated in a unsaved feature in some times).
 
There may be a virtual field "sum_missing_childern" with the expression 
`case when coalesce(relation_aggregate( relation:='myRelation', aggregate:='count',expression:="id"),0)=0 then 1 else 0 end`
and a constraint "`sum_missing_childern=0`" which enforces, that there is minimum 1 child added to each feature)

In my case one feature has no childs (1 missing child):
![grafik](https://user-images.githubusercontent.com/28433436/106150835-086afb80-617c-11eb-8ab8-e6d9fadb4179.png)
 
If I add a child, close the dialog and open the attribute table again, you can see, everything worked well: The count of missing fields is 0 now.

![grafik](https://user-images.githubusercontent.com/28433436/106151320-84fdda00-617c-11eb-8c55-be8d0b3e83df.png)

But as far as I know, virtual fields are designd to update on the fly. But after capturing a new child-feature the virtual parent-field remains unchanged ("1") although it should give a "0"-result -when the form is not closed in between and opend again

![grafik](https://user-images.githubusercontent.com/28433436/106164835-b5e50b80-618a-11eb-90ba-73ad9c06476d.png)

Virtual fields shall/can be used for constraint-purposes to get good data-quality: If you use the  virtual field `sum_missing_children` and check the enforce-expression constraint-checkbox  with a constraint "`sum_missing_childern=0`" no parent feature data will be captured without a child-feature (there are limitations because of #40129 now)

![grafik](https://user-images.githubusercontent.com/28433436/106153525-ede65180-617e-11eb-85b9-b72f09e458bf.png)

In a similar case you can use a virtual field "sum_contraint_violations_of_child_features" (as described in in #41239) with the expression 
" relation_aggregate(relation:= 'myRelation', aggregate:='sum', expression:="mistakes") 
and a constraint `sum_contraint_violations_of_child_features=0` which enforces that there are no constraint-violations in the childs you will only be able to capture valid child data too.

After adding a child or changing child-data you may expect, that the value of the virtual field changes after updating the child-feature. But this does not happen:
There is no invalid data any longer  
![grafik](https://user-images.githubusercontent.com/28433436/106154261-b330e900-617f-11eb-9527-903ab6f8268b.png)
but the sum of contraint-violations is still "2" althrough it is "0" in reality:
![grafik](https://user-images.githubusercontent.com/28433436/106154583-0c991800-6180-11eb-83bd-fae01f2a3b7c.png)
You can exit the parent-input-form only by "Cancel". If child-data is saved and you press "identify" again the virtual field is calculated correct. 

It seems that there is no trigger/signal, which forces QGIS to update the virtual field(s) of the parent-feature after changing data of the child-feature.

Sample-project is attached:
[issue_with_not_updating_virtual_fields.zip](https://github.com/qgis/QGIS/files/5888402/issue_with_not_updating_virtual_fields.zip)

**QGIS and OS versions**

QGIS version | 3.16.3-Hannover | QGIS code revision | 94ac9f21b8
-- | -- | -- | --
Compiled against Qt | 5.11.2 | Running against Qt | 5.11.2
Compiled against GDAL/OGR | 3.1.4 | Running against GDAL/OGR | 3.1.4
Compiled against GEOS | 3.8.1-CAPI-1.13.3 | Running against GEOS | 3.8.1-CAPI-1.13.3
Compiled against SQLite | 3.29.0 | Running against SQLite | 3.29.0
PostgreSQL Client Version | 11.5 | SpatiaLite Version | 4.3.0
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.8
Compiled against PROJ | 6.3.2 | Running against PROJ | Rel. 6.3.2, May 1st, 2020
OS Version | Windows 10 (10.0)
Active python plugins | changeDataSource; qfieldsync; qgis2web; quick_map_services; SpreadsheetLayers; db_manager; processing

**Comment by 3nids** [comment]
Would using layer dependencies an acceptable solution?

I am not sure we are really able to create this dependency automatically when creating the virtual field. Chances to miss something are high and it's going to be hard to maintain these dependencies based on the virtual field definition.

**Comment by koenigherford** [comment]
In a case of "missing childs" it will work because the attribute table will contain only NULL-values for non-existing childs.
In a case of "wrong data in childs" this will not work: The resulting attribute-table will just contain the data of **one** child. If there are more than 1 childs the parent-attribute-table may contain the virtual-field of a child with correct data. In this case QGIS will not realize that there is a constraint-violation in the other childs. 
I tried to create a virtual layer, which groups the childs by parent-id and adds the (virtual-) mistakes of this id.( `select a.parent_aux_id_ref, sum(a.mistakes) sum_mistakes from child_with_aux_id a group by a.parent_aux_id_ref` ). I made join between parent-layer and virtual-layer with "Cache join-layer in virtual memory" checked/unchecked state. In both times the joined field vl_sum_mistakes" did not update. Neither before saving wrong child-data nor after saving wrong child-data. Only after identifiying the parent-feature again.
The use of virtual child fields in parent-layers was an approach to have a workaround to prevent the user from saving parent tables when child-tables do not meet constraints. In my case I will do without virtual child-tables in parent tables when #41233 and #41239 are fixed. But there may be other use cases when a user may need updating virtual fields.
Sample project with virtual layer is attached.
[issue_with_not_updating_virtual_fields_virtual_layer.zip](https://github.com/qgis/QGIS/files/5902564/issue_with_not_updating_virtual_fields_virtual_layer.zip)

(Tested this in a larger project too and got a crash after a while – maybe there was a correlation to my tests)

**Comment by koenigherford** [comment]
Feedback was commented 15 days ago. If there are any further questions - let me know

**Comment by signedav** [comment]
Okay, apparently the dependencies do not work in this case. Neither on joined fields nor on fields (virtual or static) with expressions containing `relation_aggregate`. I don't know why at the moment.

**Comment by signedav** [comment]
Okay, after some more research I can summarize it as follow:

**_On having an open attribute form, field values calculated by expressions considering other layers are not updated in real time. This means when having virtual fields / fields with default values applied on update / joined fields considering values of a child layer, they will not update, while editing children in the relation editor._**

On triggering a real time refresh of a field with default values applied on update by making a change in the form itself, the data from the child layer are considered. So the information is here - just not the refresh. (Btw. I don't know how to trigger a manual refresh of the virtual fields but I guess it's just not implemented.)

So what would be needed here is a trigger of a child layer to refresh the current form. I don't know how this can be achieved automatically. Layer dependencies do not the trick - I'm not sure if they do field updates at all or if this is more to consider back-end triggers etc.

Do you have any more ideas @3nids ?

**Comment by signedav** [comment]
In the attribute form the behavior is different:
**Use Cases:**
- On changing attribute in same form -> default value fields are updated, virtual fields are not  (with or without dependencies)
- On changing other layers -> virtual fields are updated (with dependencies), default values are not (with and without depedencies)

Means dependencies are doing the trick for virtual fields in the attribute form. Only annoying thing is that it does not preserve the index, but this might be worth a (smaller) bug report:
![attrtabledependencies](https://user-images.githubusercontent.com/28384354/122022884-539e4780-cdc7-11eb-9def-8db486b5eef8.gif)

**Comment by signedav** [comment]
After all this issue here concerns these misbehaviors / not-implemented-gadgets:
- virtual fields are not updated on changing fields in the same form
- virtual fields are not updated in the attribute form when a layer depending on (with dependencies) changes
- default values "apply on update" are not updated when a layer depending on (with dependencies) changes \*
- attribute **_table_** form view looses index when a layer depending on changed

\* what is actually okay, since it should only apply updates "on edit"

**Comment by koenigherford** [comment]
Yes, it would be great to have the virtual fields calculated on the fly with their own expressions. 

This will allow to work without default values (as described in #43644 -was used as a try for a workaround) in virtual fields. In this case using virtual fields the user shall maybe prevented  to create default values by disabling "default-values"-settings-forms from QGIS-side. This will allow to close #43644.

**Comment by signedav** [comment]
This issue concerns a whole workflow. I split it now to single issues:
1. #43900
2. #43901
3. Concerning layer dependencies on default values "apply on update" -> won't fix, since it's not an issue see https://github.com/qgis/QGIS/issues/41243#issuecomment-861316615
4. #43902
